src/test/java/io/wgj/ctci/StackTest.java

```diff
@@ -2,49 +2,121 @@
         assertEquals(1, r);
     }
 
+    @Test
+    public void testStackOneString() {
+        Stack<String> s = new Stack<>();
+        s.push("one");
+        String r = s.pop();
+        assertEquals("one", r);
+    }
+
+    @Test
+    public void testPushPopsIntermixed() {
+        Stack<Integer> s = new Stack<>();
+        s.push(0);
+        assertEquals(0, (int) s.pop());
+        s.push(1);
+        s.push(2);
+        s.push(3);
+        assertEquals(3, (int) s.pop());
+        assertEquals(2, (int) s.pop());
+        s.push(4);
+        assertEquals(4, (int) s.pop());
+        assertEquals(1, (int) s.pop());
+    }
+
     @Test
     public void testStackMultipleInts() {
-        ArrayList<Integer> expected = new ArrayList<>();
-        for (int i = 2; i < 10; i = i + 2) {
-            expected.add(i);
-        }
+        Stack<Integer> s = new Stack<>();
+        s.push(1);
+        s.push(2);
+        s.push(3);
 
-        Stack s = new Stack();
-        for (int e : expected) {
-            s.push(e);
-        }
+        assertEquals(3, (int) s.pop());
+        assertEquals(2, (int) s.pop());
+        assertEquals(1, (int) s.pop());
+    }
+
+    @Test
+    public void testGrowth() {
+        Stack<Integer> s = new Stack<>(1);
+        int startingCapacity = s.capacity();
+        s.push(1);
+        s.push(2);
+        s.push(3);
+        int newCapacity = s.capacity();
+        assertTrue(startingCapacity < newCapacity);
+
+    }
 
-        Collections.reverse(expected);
-        for (int e : expected) {
-            assertEquals(e, s.pop());
+    @Test
+    public void testGrowthWithLeftOver() {
+        int cap = 257; // 257 << 2 is 1028, and trigger a "left over".
+        Stack<Integer> s = new Stack<>(cap);
+        assertEquals(cap, s.capacity());
+        for (int i = 0; i < cap + 1; i++) {
+            s.push(0);
         }
+        // Our new capacity will use what's left over from s.grow().
+        assertEquals(s.maxSize(), s.capacity());
     }
 
-    @Test(expected = ArrayIndexOutOfBoundsException.class)
-    public void testStackOverFlow() {
-        Stack s = new Stack();
-        for (int i = 0; i < 11; i++) {
-            s.push(i);
+    @Test
+    public void testMaxArraySize() {
+        Stack<Integer> s = new Stack<>(1024); // Stack.MAX_ARRAY_SIZE
+        for (int i = 0; i < s.maxSize(); i++) {
+            s.push(0);
         }
+        thrown.expect(UnsupportedOperationException.class);
+        s.push(0);
+    }
+
+    @Test
+    public void testTooLargeStartingArray() {
+        Stack<Integer> s = new Stack<>(2048);
+        assertEquals(1024, s.maxSize());
     }
 
     @Test(expected = ArrayIndexOutOfBoundsException.class)
     public void testStackUnderFlow() {
-        Stack s = new Stack();
-        s.pop();
+        new Stack().pop();
+    }
+
+    @Test
+    public void TestPeek() {
+        Stack<Integer> s = new Stack<>();
+        s.push(0);
+        assertEquals(0, (int) s.peek());
```

Comment:
I'd make these 3 tests push and pop at least 4 times to make sure they update correctly after the array resizes.

Reply:
New test `testContentsAfterGrowth()` provides similar coverage.